File: 1136_Parallel_Courses.py
from typing import List
from collections import deque

# References in nodes_in not utilized can be replaced 
# for integer in-degree for performance improvement

class Solution:
    def minimumSemesters(self, N: int, relations: List[List[int]]) -> int:
        # Base case
        if N == 1: return 1
        
        # Build an adjacency list from the relations, 
        # index by node number 1 -> N 
        nodes_out = [None]
        nodes_in = [None]
        for i in range(N): nodes_out.append(set([]))
        for i in range(N): nodes_in.append(set([]))

        for relation in relations:
            src, dst = relation
            nodes_out[src].add(dst)
            nodes_in[dst].add(src) 
        
        # Classes taken on first term and last term
        start_nodes = set([])
        end_nodes = set([])
        
        # Starting nodes have no incoming edges
        # Ending nodes have no outgoing edges
        for i in range(1, N+1):
            if len(nodes_in[i]) == 0: start_nodes.add(i)
            if len(nodes_out[i]) == 0: end_nodes.add(i)

        # Graph is not a DAG if it has no leaf nodes
        if len(end_nodes) == 0:
            return -1

        q = deque(start_nodes)
        num_sems = 0

        while q: 
            # Each "semester" the queue empties out as classes that have
            # no outstanding pre-requeisites are taken
            num_sems += 1

            # Explore only nodes present at the start of the semester
            for _ in range(len(q)):
                curr_id = q.popleft()

                # For each neighbormark each as visited by the current node 
                # (pre-req class for neighbor has been taken)
                for adj_id in nodes_out[curr_id]:
                    nodes_in[adj_id].remove(curr_id)

                    # If all pre-req classes for neighbor fufilled, add to q
                    # to be taken next semester
                    if len(nodes_in[adj_id]) == 0:
                        q.append(adj_id)
        
        # If a course was left uncompleted (its pre-reqs unfifilled)
        for i in range(1, N+1):
            if len(nodes_in[i]) > 0:
                return -1

        # All courses were completed
        return num_sems        


# Expected: 2
N = 3
relations = [[1,3],[2,3]]


# Expected: -1
# Graph creates a cycle
N = 3
relations = [[1,2],[2,3],[3,1]]


# Starting nodes have outgoing edges but no incoming
# Ending nodes have incoming edges but no outgoing
N = 9
relations = [
    [1,5], [2,5], [3,6], [3,7], [4,7],
    [5,6], [5,8], [6,8], [7,8],
    [8,9], [7,9]
]

s = Solution()
print(s.minimumSemesters(N, relations))

File: 1136_Parallel_Courses.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;

class Solution {
    public int minimumSemesters(int N, int[][] relations) {
        // Base case
        if (N == 1) { return 1; }

        // Initialize list of sets, where each index corresponds to the set of adj nodes
        ArrayList<HashSet<Integer>> node_out = new ArrayList<HashSet<Integer>>(N+1);
        node_out.add(null);
        for(int i = 0; i < N; i++){ node_out.add(new HashSet<Integer>()); }
        
        // Build adjacency list for each node, indexed by node_id
        int src, dst;
        int[] in_degree = new int[N+1];

        for(int[] rel : relations){
            src = rel[0];
            dst = rel[1];
            node_out.get(src).add(dst);
            in_degree[dst]++;
        }
        
        // Start nodes have in degree of 0
        // End nodes have out degree of 0
        ArrayList<Integer> start_nodes = new ArrayList<Integer>();
        ArrayList<Integer> end_nodes = new ArrayList<Integer>();

        for(int i = 1; i <= N; i++){
            if (in_degree[i] == 0) { start_nodes.add(i); }
            if (node_out.get(i).size() == 0) { end_nodes.add(i); } 
        }

        // A DAG must have an end node
        if (end_nodes.size() == 0){ return -1; }

        // Run topological sorting
        LinkedList<Integer> q = new LinkedList<Integer>(start_nodes);
        HashSet<Integer> curr_adj;
        int stop, curr_id, not_visited_size;
        int num_semesters = 0;

        // Continue while there are still valid classes
        while (!q.isEmpty()){
            num_semesters++;

            // Iterate through each "class" taken this semester
            stop = q.size();
            for(int i = 0; i < stop; i++){
                curr_id = q.removeFirst();
                curr_adj = node_out.get(curr_id);

                // For each class that the current class is pre-req,
                // fufill the pre-req, if all fufilled, add class to
                // queue to be taken next semester
                for(int adj_id : curr_adj){
                    not_visited_size = --in_degree[adj_id];
                    if (not_visited_size == 0) { q.addLast(adj_id); }
                }
            }
        }
        
        // Check if each class had its requirements fufilled
        // (implying that all classes were able to be taken)
        for (int num_remaining : in_degree){
            if(num_remaining > 0){ return -1; }
        }

        return num_semesters;
    }

    public static void main(String[] args){
        Solution s = new Solution();

        // Expected: 2
        // int N = 3;
        // int[][] relations = {{1,3},{2,3}};
        
        
        // Expected: -1
        // Graph creates a cycle
        // int N = 3;
        // int[][] relations = {{1,2},{2,3},{3,1}};
        
        // Expected: 5        
        // Starting nodes have outgoing edges but no incoming
        // Ending nodes have incoming edges but no outgoing
        int N = 9; 
        int[][] relations = {
            {1,5}, {2,5}, {3,6}, {3,7}, {4,7},
            {5,6}, {5,8}, {6,8}, {7,8},
            {8,9}, {7,9}
        };

        System.out.println(s.minimumSemesters(N, relations));
    }
}
